### ISSUE TITLE
Enhancement: Optional Caching of upload/downloads to improve stream stability

### ISSUE THREAD

**Issue by calisro** [ncw/rclone#711]
I would like to propose a optional "cache on copy" when using mounts and copying(or cat'ing). 

i.e. Designate a temp directory and when a copy (or stream/seek/cat) occurs cache the file at full (or bandwidth throttled) speed to the temp directory and keep serving the requests for a period of time from that location. Trying to eliminate intermittent problems from the streaming client.

Caching the uploads would allow the size of it to be known in advance and retries to be done which would make it much more reliable. Caching uploads and downloads would make them both seekable. 

The idea here is for applications that would use rclone mounts, you'd want to cache as much data as possible to the temp directory and serve that data to the client when requested.  You'd want to cache _ahead_ of the client as much as possible so that each 'seek' of the client is directed at the local cached copy.

This would require some ability to manage the size of the cache.  This could be done either by time or by 'pressure' on the cache.  i.e. define a max size and when the temp area feels pressure, it would free the oldest items if they were not actively in use.  If there was nothing to free, it would simply not cache and warn the user of the temp size may not be adequate.  We should also consider a configurable max time an item could remain in the cache to prevent very old files from being served to the client.

**Comment by ncw** [comment]
This could help with #707.

**Comment by talisto** [comment]
Maybe adding in support for [CacheFS](https://en.wikipedia.org/wiki/CacheFS) would work?

**Comment by 15siegela** [comment]
This seems to have some sort of cache system, just tried and it certainly seems to help seeking and streaming.
https://github.com/matteoserva/MegaFuse

**Comment by talisto** [comment]
I've been playing with this CacheFS FUSE project, which seems to be working really well for me:

https://github.com/cfoucher/CacheFS

It's smart enough to cache blocks, so it doesn't need to download the entire file.  Seeking still works fine.  I haven't experienced any problems with it so far.

**Comment by calisro** [comment]
Is this helping you with buffering the file? The ask here is more for the buffering side so that when it's streaming if there is a lag it has the file and a buffer downloaded ahead of time

**Comment by aus** [comment]
@ncw any thoughts on whether this feature is still being considered? I think there is a strong use-case here for people utilizing cloud services for media streaming, and I don't want this to fall off the radar.

I know a lot of folks hack a local cache together using union-type FUSE files systems like mergerfs, unionfs, or aufs. You can essentially merge a local filesystem with a similarly structured `rclone mount`. You can configure these union filesystems to give read or write priority to the local filesystem. 

Personally, I use acd_cli + encfs + unionfs (to merge ACD and local filesystems) + SnapRAID (for parity and local disk pooling). Writes always go to the local pooled filesystem. Then a cron uploads new cache files to the cloud via `acd_cli`. On good copy, it will delete files older than X days. While this works ok, there's a lot of moving pieces that often break. 

I've been testing rclone for months in my "dev" environment and it seems to be much more stable. It would be really great to have an all-in-one solution to handle local caching, encryption, mounting, and uploading. We are almost there. Caching is the last piece of the puzzle!

Lastly, I know development takes a lot of time and effort. I'd be happy to donate via a [tip4commit](https://tip4commit.com/) or something.  Lots of us owe you a few 🍻 !

**Comment by ladude626** [comment]
@talisto Can you show us how you implement this with rclone?

**Comment by talisto** [comment]
@scoopydude2002 Unfortunately CacheFS doesn't seem to be very reliable, I ran into issues that I wasn't able to resolve (crashes, incompatibility with certain filenames/pathnames, etc) so I stopped using it.  I think it's probably best to wait for a better-integrated solution.

**Comment by endiz** [comment]
Would love a caching ability! If file is read, cache file for X amount of days locally or until allocated cache is full (ex. 10gb).

**Comment by ss44** [comment]
Not sure if anyone's put any thought into this but maybe cache'ing itself could work a lot like crypt currently does. A separate share type that wraps around any existing shares.

When mounted it caches files it reads with options to specify:
- time to hold cache
- max size of cache
- size of cached block

**Comment by calisro** [comment]
Sounds like a good idea if its easier to implement.

**Comment by baro77** [comment]
Afaik it would be great if that cache worked with this logic:

1) cache should be a disk one, not a memory one; 
2) when I read,  first the cache is populated with the new data IF IT DOESN'T CONTAIN IT YET, then I get the data from the cache;
3) when I read, if the cache is ALREADY populated with the data I have requested I get the cached data EVEN IF it differs from the one on the remote;
4) when I write, I write ONLY in the cache;
5) writing-back the MODIFIED/NEW data to the remote could be done with an asyncronous "rclone sync" (and some more stuffs ;) );
6) cache cleaning should of course occur only after a sync and should be highly customizable (maybe rules-based) ... so everyone could implement by himself his preferred cache management strategy: mine for example would be to have a huge, size-only-driven cache (maybe with some files never been cleaned out of the cache too) to speed things up with frequent-used files and only daily/weekly writes-back to avoid too many, small writes to the remote

This kind of cache would implement a sort of unionfs , where the remote branch is a self-consistent baseline, the local branch is the actually used data (for everything a fs permits, non-sequential writes included), and the "out of band", asynchronous sync process represents a consolidation of local data in the the remote branch.

I don't miss these critical points: cache should handle both full files and files fragments and at each stage the data on the remote should be self-consistent (so no partial changes in the remote, neither a simple file delete, out of the full sync process of point 5) )

If you are interested in my ideas (or I haven't expressed them effectively :-( ), let me know
thanks
Andrea

**Comment by jakobrosenberg** [comment]
Ccould bcachefs be an alternative to cachefs?

https://bcache.evilpiepirate.org

**Comment by jakobrosenberg** [comment]
@talisto did you notice a pattern with the issues you had with cachefs? I'm trying it out and for some reason it's unable to work with half my folders. It's a shame, cause the caching is otherwise perfect.

**Comment by baro77** [comment]
My own case study, strictly Mac-based but I hope it can be useful for someone: I use rclone fuse as a read-only branch of unionfs 2.0 with copy on write enabled  .. of course the rw branch is a folder on my local hard disk.... you would say: "yes, but the rw branch isn't a cache, 'cause it gets populated only when you access a file in write mode..."... that's where the trick comes in... I do not directly use the unionfs not for my files, in fact for a sparsebundle (which of course I mount and use as a real Mac HFS+ formatted HD)... for non-Mac geeks out there: sparse bundle is a disk image, dinamically growing as needed, actually based on a bunch of 8MB files  (the so-called "bands"), formatted with Mac native FS (so I don't miss any peculiar Mac disks feature), and crypted by AES, if you need (which is a must for me when I use clouds). 
The "magic" is that unionfs beneath just sees 8 MB files (the bands), not the files I use after the sparsebundle is mounted... and of course there isn't a one-to-one correspondence between bands and files contained in them, so when I open a file, even if only for read access, the arbitrator opens the involved bands RW (because few instants later they could be involved in opening another contained file in RW mode).... so unionfs copies them in local folder (the RW branch) :)
Now I'm working on a script for copying my local changes to the cloud (it's a critical point, I want the online copy to always be self-consistent and usable even if not up-to-date, so when I "sync" local changes I have to be sure to copy and delete and modify all that is needed) and on a cache-freeing algorithm based on bands-usage metadata contained in sqlite

**Comment by cardinalfan1** [comment]
Here's another code-base that may be helpful: https://github.com/Doloops/mcachefs

**Comment by Lillecarl** [comment]
I really do like the idea @ss44 shared, that the cache function could be a wrapper around the rest of the filesystems. So on layers we got: cloud <- crypt <- cache.

What struck me though is, how will this work with a bittorrent application that never closes it's file handles (?). Then cache wouldn't know when to start upload.

I'd love this feature, so that i can go onto using Backblaze B2 instead of Ebay Google.

**Comment by seuffert** [comment]
I am currently looking into this issue (since i need it myself). Is there currently any work going on?

The idea of a new "caching remote" as a wrapper for another sounds suitable. My personal scenario only raises the need for a r/o caching remote, so that's what i am planing to build.

mount <=> disk-cache remote (max size, chunk size, LRU cleaning) <=> source remote

**Comment by ncw** [comment]
@seuffert I hadn't thought of making a caching remote.  I think this is only of interest to mount so my plan was to integrate it there.

I'm not working on it at the moment, though I would like to get this feature into 1.38.

I'd quite like to work on the write caching first so if you are interested in read caching then that is good!

**Comment by remusb** [comment]
I'm actually working on caching remote and the functionality is there. I just need to write the tests and overcome a nasty fuse issue on macos to allow easier testing.

**Comment by seuffert** [comment]
@remusb sounds terrific! If needed I would Iike to offer my help with testing, refining or just simply bug hunting. How is the cache designed? Chunks, clearing algorithm etc

**Comment by vampywiz17** [comment]
@remusb you work a own project (program) or a rclone fork?

**Comment by remusb** [comment]
@vampywiz17 PR in rclone so it will come in here not in a separare tool or project

@seuffert sure
Fair warning, the code is not in a presentable state yet. This was my introduction to golang and getting to know rclone's internals so I had to rewrite stuff a lot of times (that's how I learn really)
It's all in /cache and a fix in crypt that didn't allowed it to be wrapped by another remote.
Important things:
- I'm still adding some write support
- uses chunks in bolt, will likely clean by old usage but I was thinking on adding multiple types
- dir lists are also cached and cleaned by timestamps
Would love to get some feedback on the design choices. I don't personally like how I handled the dir listings so far. 

https://github.com/remusb/rclone/tree/backend-cache/cache

**Comment by ncw** [comment]
@remusb Good work :-)

I see you are serializing fs.Object~s.  I have a plan to add a couple of optional methods to Fses to allow object serialization and deserialization - I think this will make your life easier.  You then wouldn't have to call NewObject in CachedObject.RefreshObject (which can be expensive).  I could bring forwards this plan if you want.

To be truly useful with mount, mount needs to be able to do non sequential writes, and reads from the same object.  We could make two new optional interfaces on the Fs support that, Say ReadBlockAt and WriteBlockAt, or maybe on the open handle might be eaiser.

BTW your ListR implementation looks wrong - it doesn't seem to be recursing...

I suggest you make a PR at some point which I won't merge until you are ready.  That will run loads of checks on your code - some of which I can see will fail.  You can run them locally if you want (check out the contents of `.travis.yml` and run `make check`, `make quicktest`, `make compile_all`). A PR is a good way of getting my attention to look at code too!

**Comment by ncw** [comment]
I'm please to announce that after a lot of hard work by @remusb we have merged the first version of a cache backend for rclone.

Please have a go and check out [the docs](https://github.com/ncw/rclone/blob/master/docs/content/cache.md) and have a go!

Please report any problems in a [new issue on github](https://github.com/ncw/rclone/issues/new).

You'll find support in [the latest beta](https://beta.rclone.org).

I'm currently working on proper read/write support for mount which will also involve some caching - this isn't quite merged yet though.

**Comment by ncw** [comment]
I've now added the mount changes.  Here is a summary of the new flags from the docs.

The emphasis is on making the filesystem usable for more uses - the caching is a side effect.

Now that this and the cache backend are both in, I'm going to close this issue - please report bugs in new issues!

### File Caching ###

**NB** File caching is **EXPERIMENTAL** - use with care!

These flags control the file caching options.

    --cache-dir string               Directory rclone will use for caching.
    --cache-max-age duration         Max age of objects in the cache. (default 1h0m0s)
    --cache-mode string              Cache mode off|minimal|writes|full (default "off")
    --cache-poll-interval duration   Interval to poll the cache for stale objects. (default 1m0s)

If run with ` + "`-vv`" + ` rclone will print the location of the file cache.  The
files are stored in the user cache file area which is OS dependent but
can be controlled with ` + "`--cache-dir`" + ` or setting the appropriate
environment variable.

The cache has 4 different modes selected by ` + "`--cache-mode`" + `.
The higher the cache mode the more compatible rclone becomes at the
cost of using disk space.

Note that files are written back to the remote only when they are
closed so if rclone is quit or dies with open files then these won't
get written back to the remote.  However they will still be in the on
disk cache.

#### --cache-mode off ####

In this mode the cache will read directly from the remote and write
directly to the remote without caching anything on disk.

This will mean some operations are not possible

  * Files can't be opened for both read AND write
  * Files opened for write can't be seeked
  * Files open for read/write with O_TRUNC will be opened write only
  * Files open for write only will behave as if O_TRUNC was supplied
  * Open modes O_APPEND, O_TRUNC are ignored
  * If an upload fails it can't be retried

#### --cache-mode minimal ####

This is very similar to "off" except that files opened for read AND
write will be buffered to disks.  This means that files opened for
write will be a lot more compatible, but uses the minimal disk space.

These operations are not possible

  * Files opened for write only can't be seeked
  * Files open for write only will behave as if O_TRUNC was supplied
  * Files opened for write only will ignore O_APPEND, O_TRUNC
  * If an upload fails it can't be retried

#### --cache-mode writes ####

In this mode files opened for read only are still read directly from
the remote, write only and read/write files are buffered to disk
first.

This mode should support all normal file system operations.

If an upload fails it will be retried up to --low-level-retries times.

#### --cache-mode full ####

In this mode all reads and writes are buffered to and from disk.  When
a file is opened for read it will be downloaded in its entirety first.

In this mode, unlike the others, when a file is written to the disk,
it will be kept on the disk after it is written to the remote.  It
will be purged on a schedule according to ` + "`--cache-max-age`" + `.

This mode should support all normal file system operations.

If an upload or download fails it will be retried up to
--low-level-retries times.

**Comment by ncw** [comment]
NB in f80f7a050948d40f67edeb93f6757d30ec85d3ed I've changed the default location of the cache so you might need to move your cache, or start again if you aren't using the `--cache-db-path`.  This was to move it to a windows/mac/linux approved place like .cache.

Sorry for the inconvenience, but I thought it was better to do this now!

**Comment by Saren-Arterius** [comment]
I made a 2GB image inside gsuite rclone mount, with `--vfs-cache-mode=full`, I tried to format the image and mount it for science, this is the result

- `mkfs.btrfstest.img`: `No valid Btrfs found on /dev/sdc ERROR: open ctree failed. `
- `mkfs.ext4 test.img`: Success
- `sudo mount test.img /mnt`: `mount: /mnt: mount failed: Operation not permitted.`

I wonder why operation not permitted?
